sdk/src/env.rs

```diff
@@ -254,6 +254,89 @@ impl Env {
     pub fn invoke_contract(&mut self, hf: xdr::HostFunction, args: xdr::ScVec) -> xdr::ScVal {
         self.env_impl.invoke_function(hf, args).unwrap()
     }
+
+    #[cfg(not(target_family = "wasm"))]
+    fn clone_self_and_catch_panic<F, T>(&self, f: F) -> (Env, std::thread::Result<T>)
+    where
+        F: FnOnce(Env) -> T,
+    {
+        let hook = std::panic::take_hook();
+        std::panic::set_hook(Box::new(|_| ()));
+        let deep_clone = self.deep_clone();
+        let res = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| f(deep_clone.clone())));
+        std::panic::set_hook(hook);
+        (deep_clone, res)
+    }
+
+    #[cfg(not(target_family = "wasm"))]
+    pub fn assert_panic_with_string<F, T: Debug>(&self, s: &str, f: F)
```

Comment:
💡 I think we're making more assumptions than we need to about how developers will want to assert on panic / status errors by providing functions that retrieve the value and assert on it. I think we could defer the decision to developers on how they will assert on the error by simply returning the error to them in a `Result`.

This will have two advantages:

- For tests that have a return value as a Result, and for calls that the developer doesn't want to fail, they can simply chuck a `?` on the end and they can surface a `Status` error.
   ```rust
   let e = Env::with_empty_recording_storage();
   let got = e.capture_status(|env| ...)?;
   ```

- For developers that want to assert more complex things than a simple equality, they can do that however they like. And equality checks will still be easy:
   ```rust
   let e = Env::with_empty_recording_storage();
   let got: Result<_, Status> = e.capture_status(|env| ...);
   assert_eq!(got, Err("panic string here"))
   ```

I think both examples above work with strings and statuses. And then we can apply the same pattern to the `Env::invoke_contract` functions so that whenever you invoke a contract you get back a `Result<_, Status>`.

cc @graydon @jonjove